sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala

```diff
@@ -124,12 +124,45 @@ case class GroupExpression(children: Seq[Expression]) extends Expression {
   override def dataType: DataType = throw new UnsupportedOperationException
 }
 
+trait TypeConstraint {
+  self: Expression =>
+
+  def isTypeMatch: Boolean
+
+  def errorMessage: String
```

Comment:
this cannot be just called errorMessage anymore once you put it in Expression. need something more descriptive.